### PR TITLE
adr-tools 1.2.0 (new formula)

### PR DESCRIPTION
**Pull request by jhulten**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally prior to submission with `brew install <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [x] Does your submission pass `brew audit --new-formula <formula>` (after doing `brew install <formula>`)?

---

Add adr-tools from https://github.com/npryce/adr-tools/ to support
managing architecture decision records.

### CODE REVIEW

File: Formula/adr-tools.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      s.gsub! 'adr_template_dir="$adr_bin_dir"', "adr_template_dir=\"#{prefix}\""
+    end
+
+    inreplace Dir["src/adr-*"], 'source "$(dirname $0)/config.sh"', "source '#{prefix}/config.sh'"
+    inreplace Dir["src/_adr_*"], 'source "$(dirname $0)/config.sh"', "source '#{prefix}/config.sh'"
+
+    inreplace "src/adr" do |s|
+      s.gsub! 'source "$(dirname $0)/config.sh"', "source '#{prefix}/config.sh'"
+      s.gsub! "cmd=$(dirname $0)/adr-$1", "cmd=$adr_bin_dir/adr-$1"
+    end
```
Comment (MikeMcQuaid) [inline Formula/adr-tools.rb]: Can you please submit these patches to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

File: Formula/adr-tools.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    prefix.install Dir["src/*.md"]
+    prefix.install Dir["src/config.sh"]
+    bin.install Dir["src/*"] - Dir["src/*.md"] - Dir["src/config.sh"]
```
Comment (MikeMcQuaid) [inline Formula/adr-tools.rb]: If you've `.install`ed above those files will no longer exist.

File: Formula/adr-tools.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    inreplace Dir["src/adr-*"], 'source "$(dirname $0)/config.sh"', "source '#{prefix}/config.sh'"
+    inreplace Dir["src/_adr_*"], 'source "$(dirname $0)/config.sh"', "source '#{prefix}/config.sh'"
+
+    inreplace "src/adr" do |s|
+      s.gsub! 'source "$(dirname $0)/config.sh"', "source '#{prefix}/config.sh'"
+      s.gsub! "cmd=$(dirname $0)/adr-$1", "cmd=$adr_bin_dir/adr-$1"
+    end
+
+    prefix.install Dir["src/*.md"]
+    prefix.install Dir["src/config.sh"]
```
Comment (MikeMcQuaid) [inline Formula/adr-tools.rb]: Should this be in `etc`?

File: Formula/adr-tools.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    bin.install Dir["src/*"] - Dir["src/*.md"] - Dir["src/config.sh"]
+  end
+
+  test do
+    system "true"
```
Comment (MikeMcQuaid) [inline Formula/adr-tools.rb]: This needs a test.

File: Formula/adr-tools.rb (hunk 3 above)
Comment (MikeMcQuaid) [inline Formula/adr-tools.rb]: Also: https://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/6397/version=el_capitan/testReport/junit/brew-test-bot/el_capitan/audit_adr_tools___new_formula/

File: Formula/adr-tools.rb (hunk 1 above)
Comment (jhulten) [inline Formula/adr-tools.rb]: So line 19 is viably an upstream fix. The developer currently says 'check out the directory and add it to your path' but I can push for better layout and proper install if that is needed.

Comment (MikeMcQuaid) [inline Formula/adr-tools.rb]: We need to figure out a long-term way to get this working without patching. Another option would be to use symlinks or `bin.write_exec_script`.

**Comment by dunn** [comment]
Any update here?

File: Formula/adr-tools.rb (hunk 3 above)
Comment (jhulten) [inline Formula/adr-tools.rb]: Unless I `gsub!` the destinations as well, the code looks for `config.sh`. This isn't ideal, but again it comes down to how much change @npryce is willing to take on his end (assuming I ship some PRs his way).

File: Formula/adr-tools.rb (hunk 1 above)
Comment (MikeMcQuaid) [inline Formula/adr-tools.rb]: Nudge on this.

Comment (jhulten) [inline Formula/adr-tools.rb]: RE: L18

How would you approach this, @MikeMcQuaid? While the original developer expected `config.sh` to exist in the same directory as the adr commands, this seems like a *terrible* idea unless I rename it and then I have a different set of gsubs.

Comment (MikeMcQuaid) [inline Formula/adr-tools.rb]: > While the original developer expected config.sh to exist in the same directory as the adr commands

That's fine; install everything to `libexec` and use `bin.write_exec_script Dir["#{libexec}/bin/*"]`

Comment (MikeMcQuaid) [inline Formula/adr-tools.rb]: @jhulten Any news?

**Comment by npryce** [comment]
Hi. I'm the developer of adr-tools. Happy to change the way configuration works. I added config.sh to try and avoid the need to edit the source like this when packaging the app. But it's obviously not a good mechanism. Please raise issue(s) on the adr-tools project.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by MikeMcQuaid** [comment]
Passing on this, I'm afraid, until upstream are able to provide a more flexible configuration mechanism or `bin.write_exec_script Dir["#{libexec}/bin/*"]` is working on a new PR. Sorry and thanks anyway!

**Comment by npryce** [comment]
If someone tells me what you need, I can implement it in the upstream repo.  But so far I've only received a link to this thread, and no request for any changes.

**Comment by MikeMcQuaid** [comment]
@npryce My apologies, I wasn't expecting you to do anything but @jhulten or another user to raise the issue. My take is removing our `inreplace` calls and providing similar support to them with a `configure` script or similar. Thanks!

**Comment by jhulten** [comment]
And I have been hip deep in preparing a house to move into, so my non-work geek time is limited. 

@npryce The issue I see is that the config file is sourced and expected to be in the same directory as the executables. If that was converted to a command `adr-env` and loaded defaults, plus overrides and presented values as environment variables (ala `docker-machine env` or whatnot) that would allow the other components to `eval $(adr-env)` instead of `source "$(dirname $0)/config.sh"`.

I will revisit this when I have mental space on the brew packaging end. No need to reopen.

**Comment by npryce** [comment]
I've added a command `adr-config` that outputs what was the contents of config.sh, and made the adr scripts eval the output.  This will, I hope, let packagers configure things by writing their own adr-config scripts that output variables that reflect how they have packaged the application.

Let me know if that works or if it needs any changes.

**Comment by MikeMcQuaid** [comment]
@npryce Thanks. That's an improvement but ideally there'd be a `configure`-style script on your end which does the current replacements for us (even if you're literally just calling `sed` with the same stuff). Our main reasoning for that is that you know your own project better than we ever will and if you're making some sweeping changes in a future version that require changing or tweaking those `sed` lines then you'll be in a better position to do so than a random Homebrew contributor who just wants the latest version of your stuff. Thanks for being a fanasticly responsive upstream maintainer regardless, though! 🎉

**Comment by npryce** [comment]
I'm not sure I understand what you want.  A configure script that changes the variable definitions in adr-config?  Or a configure script _instead_ of adr-config that rewrites all the source to set variables in them?  

The latter sounds too risky. 

 The former doesn't sound it would help -- if I had to change the structure of the program(s), I'd have to change the variables, which means a configure script would have different command-line parameters, and you'd have to change your packaging scripts.  Instead, I expect packagers to have their own adr-config script and replace the one in the source when building the package.

**Comment by MikeMcQuaid** [comment]
@npryce I want to be able to do `./configure --prefix=#{prefix}` or a Makefile or CMake equivalent and then `make install`. I don't want us to have to supply our own scripts or (as currently) perform replacements of other files, I'd like us to be able to specify an installation prefix and that is all. Note that that's the most standard way of installing most built-from-source software. How that's actually implemented under the hood I have no real preference.

**Comment by kevinjalbert** [comment]
Looking at the recent [ADR](https://github.com/npryce/adr-tools/blob/bd5c858c8bf33dde90d41d30d2e36528881d2762/doc/adr/0007-invoke-adr-config-executable-to-get-configuration.md) it seems like the config issue _should_ be resolved as of [2.0.0](https://github.com/npryce/adr-tools/releases/tag/2.0.0)?

> Added a mechanism for packagers to configure the application to match the installation conventions of the target platform

Is that the case? Thus allowing movement on the inclusion of it in homebrew?

**Comment by npryce** [comment]
It could be packaged for Homebrew (and other software repositories) already. The change you refer to made it easier to do so.

**Pull request closed without merging**